### PR TITLE
Better error messaging and some docs improvements

### PR DESCRIPTION
**Pull request by leopd**
A few changes to error messaging.  

In several places, python errors were just printing the string message of the exception, not the exception class or stack-trace.  Now it's printing the full stack-trace.  This is mostly in python operators.

In a couple of the C++ operators I've improved the error messages when checks fail.

Also a few comment changes which will improve the auto-generated API docs.

### CODE REVIEW

File: python/mxnet/test_utils.py [hunk 1]
```diff
@@ -688,7 +689,7 @@ def check_consistency(sym, ctx_list, scale=1.0, grad_req='write'):
             try:
                 npt.assert_allclose(arr1, arr2, rtol=tol[dtypes[i]], atol=tol[dtypes[i]])
             except Exception as e:
-                print(e)
+                traceback.print_exc()
```
Comment (piiswrong) [inline python/mxnet/test_utils.py]: this should actually rethrow. I thought I fixed it

Comment (leopd) [inline python/mxnet/test_utils.py]: Okay, added `raise` to re-throw both of these.

**Comment by piiswrong** [comment]
Please fix lint

**Comment by leopd** [comment]
How can I see the linter errors?  The link to the failed test http://54.84.234.35:8080/job/mxnet/399/ times out for me.

**Comment by leopd** [comment]
Figured out my networking issue accessing the Jenkins server.

**Comment by leopd** [comment]
I think the only lint errors remaining are the same 14 warnings that were there before.

**Comment by piiswrong** [comment]
http://107.23.36.180:8080/job/mxnet/357/console
This page shows the lint error:

21:55:37 =====134/135 cpp-header files passed check=====
21:55:37 src/operator/pooling-inl.h: 2 Errors of 1 Categories map={'whitespace': 2}
21:55:37 =====147/147 cpp-soruce files passed check=====
21:55:37 =====39/41 python files passed check=====
21:55:37 python/mxnet/io.py: 1 Errors of 1 Categories map={'convention': 1}
21:55:37 python/mxnet/operator.py: 2 Errors of 1 Categories map={'convention': 2}
21:55:37 3 files failed lint

**Comment by piiswrong** [comment]
also please rebase to resolve conflict

**Comment by piiswrong** [comment]
lint error depends on verion
you need to install this version of lint:
pip install cpplint 'pylint==1.4.4' 'astroid==1.3.6'

**Pull request merged**